src/utils/cosmosdb_model_versioned.ts

```diff
@@ -0,0 +1,190 @@
+  version: NonNegativeInteger
+});
+
+export type VersionedModel = t.TypeOf<typeof VersionedModel>;
+
+/**
+ * Returns a string with a composite id that has the format:
+ * MODEL_ID-VERSION
+ *
+ * MODEL_ID is the base model ID
+ * VERSION is the zero-padded version of the model
+ *
+ * @param modelId The base model ID
+ * @param version The version of the model
+ */
+export function generateVersionedModelId(
+  modelId: ModelId,
+  version: NonNegativeInteger
+): string {
+  const paddingLength = 16; // length of Number.MAX_SAFE_INTEGER == 9007199254740991
+  const paddedVersion = ("0".repeat(paddingLength) + String(version)).slice(
+    -paddingLength
+  );
+  return `${modelId}-${paddedVersion}`;
+}
+
+const incVersion = (version: NonNegativeInteger) =>
+  (Number(version) + 1) as NonNegativeInteger;
+
+/**
+ * Assumption: the model ID is also the partition key
+ */
+export abstract class CosmosdbModelVersioned<
+  T,
+  TN extends Readonly<T & Partial<VersionedModel>>,
+  TR extends Readonly<T & VersionedModel & BaseModel>
+> extends CosmosdbModel<T, TN & BaseModel, TR> {
+  constructor(
+    container: Container,
+    protected readonly newVersionedItemT: t.Type<TN, ItemDefinition, unknown>,
+    protected readonly retrievedItemT: t.Type<TR, unknown, unknown>,
+    protected readonly modelIdKey: keyof T
+  ) {
+    super(
+      container,
+      t.intersection([newVersionedItemT, BaseModel]),
+      retrievedItemT
+    );
+  }
+
+  public create = (
+    o: TN,
+    options?: RequestOptions
+  ): TaskEither<CosmosErrors, TR> => {
+    // if we get an explicit version number from the new document we use that,
+    // or else we set version to 0
+    const currentVersion: NonNegativeInteger | undefined = o.version;
+    const version =
+      currentVersion === undefined ? (0 as NonNegativeInteger) : currentVersion;
+
+    // the ID of each document version is composed of the document ID and its version
+    // this makes it possible to detect conflicting updates (concurrent creation of
+    // profiles with the same profile ID and version)
+    const modelId = this.getModelId(o);
+    const versionedModelId = generateVersionedModelId(modelId, version);
+
+    const newDocument = {
+      ...o,
+      id: versionedModelId,
+      version
+    } as TN & VersionedModel & BaseModel;
+
+    return super.create(newDocument, options);
+  };
+
+  /**
+   * Creates a new version from a full item definition
+   */
+  public upsert = (o: TN): TaskEither<CosmosErrors, TR> => {
+    // if we get an explicit version number from the new document we use that,
+    // or else we get the last version by querying the database
+    const currentVersion: NonNegativeInteger | undefined = o.version;
+    const modelId = this.getModelId(o);
+    return (currentVersion === undefined
+      ? this.getNextVersion(modelId)
+      : fromEither<CosmosErrors, NonNegativeInteger>(right(currentVersion))
+    ).chain(nextVersion =>
```

Comment:
Ensure is NonNegativeInteger

```suggestion
    return NonNegativeInteger.decode(currentVersion)
      .fold(
        _ => this.getNextVersion(modelId),
        _ => taskEither.of(_)
      )
      .chain(nextVersion =>
```